frontends/common/constantFolding.h

```diff
@@ -23,36 +23,97 @@ limitations under the License.
 #include "frontends/p4/typeChecking/typeChecker.h"
 
 namespace P4 {
-// Can also be run before type checking, so it only uses types if
-// they are available.
+
+/** Implements a pass that statically evaluates many constant
+ * expressions.
+ *
+ * Currently the following expressions are treated as constants:
+ *  - `IR::Constant`
+ *  - `IR::BoolLiteral`
+ *  - `IR::ListExpression` where each element is also a constant
+ *  - `IR::TypeNameExpression`
```

Comment:
This also handles constants of enum types. Perhaps it should also handle constants of the error type...